AP_HAL_ChibiOS: Added RADIX2HD

mluessi opened the pull request:
Adds support for the BrainFPV RADIX 2 HD flight controller. 

The target is H750 based. This code depends on #23947

andyp1per commented:
So is the bootloader closed source?

mluessi commented:
> So is the bootloader closed source?

Yes, it's closed source. But we have an open source tool to pack firmware for it: [brainfpv_fw_packer](https://github.com/BrainFPV/brainfpv_fw_packer)

mluessi commented:
@andyp1per Now that the other PR has been merged, this should be ready. Please let me know if there is anything else needed.

andyp1per commented:
Needs a rebase and the tests passing

mluessi commented:
@andyp1per rebased and squashed commits

tridge reviewed (changes requested):
please run:
Tools/gittools/git-subsystems-split

mluessi commented:
@tridge thanks for the review!

I added the readme file and split the commits. I also added an example script that shows how to use the firmware packer tool.

mluessi commented:
Please let me know if anything else is needed. If it can be merged, I will rebase it again to fix the board type conflict.

tridge commented:
@mluessi needs rebase and new board ID

tridge commented:
@andyp1per why dev call without approval?

peterbarker commented:
> Please let me know if anything else is needed. If it can be merged, I will rebase it again to fix the board type conflict.

We suggest making a separate PR to reserve your board ID - we merge those quickly to avoid conflicts on these longer-lived PRs.

tridge commented on libraries/AP_HAL_ChibiOS/hwdef/RADIX2HD/pack_firmware.sh:
```diff
@@ -0,0 +1,28 @@
+
+# Edit these paths to be correct
+PACKER=/shared/brainfpv/brainfpv_code/brainfpv_fw_packer/brainfpv_fw_packer/brainfpv_fw_packer.py
+PATH=/shared/brainfpv/code/ardupilot/tools/gcc-arm-none-eabi-6-2017-q2-update/bin/:${PATH}
```
what is this script for? it seems to have personal paths? is it needed?

mluessi replied:
The target uses a proprietary bootloader and needs a firmware file that is in a custom file format. The file can be created using the [BrainFPV Firmare  Packer](https://github.com/BrainFPV/brainfpv_fw_packer) script, which is available open source.

I fixed the `pack_firmware.sh` script such that it no longer uses personal paths.

tridge commented on libraries/AP_HAL_ChibiOS/hwdef/RADIX2HD/README.md:
```diff
@@ -0,0 +1,98 @@
+Note: When using Ardupilot, it is necessary to have a microSD card inserted, without it
+the firmware won't run.
+
+To create custom binaries, use the [BrainFPV Firmware Packer](https://github.com/BrainFPV/brainfpv_fw_packer)
```
what is a "custom binary" ?

mluessi replied:
That was poorly worded. It's just a firmware file in a special format (see above). I updated the readme and added an example for how to build the Copter firmware and use the script to create a firmware file for the RADIX2HD

andyp1per commented:
> @andyp1per why dev call without approval?

Because I was really not sure what the result should - marked it for devcall so that it could be discussed.

mluessi commented:
> > Please let me know if anything else is needed. If it can be merged, I will rebase it again to fix the board type conflict.
> 
> We suggest making a separate PR to reserve your board ID - we merge those quickly to avoid conflicts on these longer-lived PRs.

Yes, this makes sense. I made a separate PR, see #24152

andyp1per commented on libraries/AP_HAL_ChibiOS/hwdef/RADIX2HD/hwdef.dat:
```diff
@@ -0,0 +1,187 @@
+MCU STM32H7xx STM32H750xx
+
+# board ID for firmware load
+APJ_BOARD_ID 1117
```
Needs a new board id

mluessi replied:
I changed it to 1118 as in #24152

andyp1per commented on libraries/AP_HAL_ChibiOS/hwdef/RADIX2HD/hwdef.dat:
```diff
@@ -0,0 +1,187 @@
+
+# USART2 (GPS)
+PD6 USART2_RX USART2
+PD5 USART2_TX USART2
```
Think you will need to define the GPS protocol here as the default protocol is mavlink for SERIAL2

mluessi replied:
Thanks, I changed it to GPS

andyp1per commented on libraries/AP_HAL_ChibiOS/hwdef/RADIX2HD/hwdef.dat:
```diff
@@ -0,0 +1,187 @@
+
+# UART4 (ESC telemetry)
+PB8 UART4_RX UART4 NODMA
+PA0 UART4_TX UART4 NODMA
```
Can define this as ESC telemetry protocol

mluessi replied:
I didn't set it to default to ESC telemetry. Some users may want it and the RX4 signal is available on the 4-in-1 ESC connector, but it isn't typically used in my experience.

andyp1per commented on libraries/AP_HAL_ChibiOS/hwdef/RADIX2HD/hwdef.dat:
```diff
@@ -0,0 +1,187 @@
+PD14 TIM4_CH3 TIM4 PWM(8)  GPIO(57) BIDIR # M8
+
+# PWM output on UART6 TX and RX pins
+PC6 TIM8_CH1 TIM8 PWM(9)  GPIO(58)        # TX6
```
If these are only PWM you can specify NODMA to save channels

mluessi replied:
Yes, good idea. I added NODMA to those

andyp1per commented on libraries/AP_HAL_ChibiOS/hwdef/RADIX2HD/hwdef.dat:
```diff
@@ -0,0 +1,187 @@
+define AP_NOTIFY_TONEALARM_ENABLED 1
+
+# printf on USART2
+STDOUT_SERIAL SD2
```
This will prevent debug messages doing to the console, I think you want to remove this

mluessi replied:
I remove it

andyp1per commented on libraries/AP_HAL_ChibiOS/hwdef/RADIX2HD/hwdef.dat:
```diff
@@ -0,0 +1,187 @@
+define HAL_COMPASS_DEFAULT HAL_COMPASS_NONE
+define HAL_I2C_INTERNAL_MASK 0
+
+define BOARD_PWM_COUNT_DEFAULT 8
```
I didn't think we needed to define this anymore

mluessi replied:
Removed those as well

andyp1per reviewed (changes requested):
Some fixups required

mluessi commented:
@andyp1per thank you for the review and the comments. I made the requested changes. Please let me know if you have any other input.

andyp1per commented:
@mluessi just need to squash the commits together and then I think you are good to go

mluessi commented:
> @mluessi just need to squash the commits together and then I think you are good to go

Great :). Commits squashed
